File: backend/src/notifications/telegram.rs
use std::sync::Arc;
use crate::{app::AppState, db::queries, pipeline::trigger::get_setting_or_env};

pub enum TelegramEvent {
    JobQueued(String),
    JobStarted(String),
    CheckpointSaved(String, String),
    JobCompleted(String, String),
    JobFailed(String, String, String),
}

pub async fn send_telegram_notification(state: Arc<AppState>, event: TelegramEvent) {
    // Check if notifications enabled
    let enabled = queries::get_setting(&state.db, "notifications_enabled").await
        .ok()
        .flatten()
        .map(|v| v == "true")
        .unwrap_or(false);
    if !enabled { return; }

    let bot_token = match get_setting_or_env(&state, "telegram_bot_token").await {
        Ok(Some(t)) => t,
        _ => return,
    };
    let channel_id = match get_setting_or_env(&state, "telegram_channel_id").await {
        Ok(Some(c)) => c,
        _ => return,
    };

    let message = format_telegram_message(&event);

    let url = format!("https://api.telegram.org/bot{}/sendMessage", bot_token);
    let _ = state.http.post(&url)
        .json(&serde_json::json!({
            "chat_id": channel_id,
            "text": message,
            "parse_mode": "HTML"
        }))
        .send()
        .await;
}

fn format_telegram_message(event: &TelegramEvent) -> String {
    match event {
        TelegramEvent::JobQueued(title) => {
            format!("🎬 <b>Added to queue:</b> {}", title)
        }
        TelegramEvent::JobStarted(title) => {
            format!("⚙️ <b>Processing started:</b> {}", title)
        }
        TelegramEvent::CheckpointSaved(title, phase) => {
            format!("💾 <b>Checkpoint saved:</b> {} — {}", title, phase)
        }
        TelegramEvent::JobCompleted(title, details) => {
            format!(
                "━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n✅ <b>StreamVault - Download Complete</b>\n\n🎬 {}\n{}",
                title, details
            )
        }
        TelegramEvent::JobFailed(title, phase, error) => {
            format!("❌ <b>Failed:</b> {} at {} — {}", title, phase, error)
        }
    }
}

File: backend-bun/src/notifications/telegram.ts
import type { Context } from "hono";
import type { DrizzleDB } from "../db/index";
import type { Config } from "../config";
import { getSetting } from "../db/queries";

// ── Types ──

export interface TelegramEvent {
  type:
    | "JobQueued"
    | "JobStarted"
    | "CheckpointSaved"
    | "JobCompleted"
    | "JobFailed";
  title: string;
  phase?: string;
  error?: string;
  details?: string;
}

// ── Globals for scheduler path (no request context) ──

let _db: DrizzleDB | null = null;
let _config: Config | null = null;

export function setNotificationGlobals(db: DrizzleDB, config: Config): void {
  _db = db;
  _config = config;
}

// ── Message formatting ──

export function formatMessage(event: TelegramEvent): string {
  switch (event.type) {
    case "JobQueued":
      return `🎬 <b>Added to queue:</b> ${event.title}`;
    case "JobStarted":
      return `⚙️ <b>Processing started:</b> ${event.title}`;
    case "CheckpointSaved":
      return `💾 <b>Checkpoint saved:</b> ${event.title} — ${event.phase}`;
    case "JobCompleted":
      return [
        "━━━━━━━━━━━━━━━━━━━━━━━━━━━━━",
        "✅ <b>StreamVault - Download Complete</b>",
        "",
        `🎬 ${event.title}`,
        event.details ?? "",
      ].join("\n");
    case "JobFailed":
      return `❌ <b>Failed:</b> ${event.title} at ${event.phase} — ${event.error}`;
  }
}

// ── Send ──

export async function sendNotification(
  c: Context<any> | null,
  event: TelegramEvent,
): Promise<void> {
  const db = c?.var?.db ?? _db;
  const config = c?.var?.config ?? _config;
  if (!db || !config) return;

  const enabled = getSetting(db, "notifications_enabled");
  if (enabled !== "true") return;

  const botToken =
    getSetting(db, "telegram_bot_token") || config.telegramBotToken;
  const channelId =
    getSetting(db, "telegram_channel_id") || config.telegramChannelId;
  if (!botToken || !channelId) return;

  const text = formatMessage(event);

  // ponytail: fire-and-forget, .catch swallows network errors silently
  fetch(
    `https://api.telegram.org/bot${botToken}/sendMessage`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        chat_id: channelId,
        text,
        parse_mode: "HTML",
      }),
    },
  ).catch((e) => console.error("[telegram] notification failed:", e));
}
